### PR TITLE
Prevent module activation when plan doesn't support them

### LINKED ISSUE #7867: SEO Tools can be enabled from module list
In 3302536-t the user was seeing duplicate meta description tags on his post. One from Yoast, and another which disappeared when Jetpack was deactivated.

The debug tool showed that SEO Tools was enabled even though the user has never had a Professional Plan.

After some digging, I found that even without a Professional Plan, you can activate the SEO Tools module from the full modules list, which causes meta descriptions to be output in posts.

Should that module only be in the list if you have a Pro Plan? (Or show an upgrade prompt there?)

### PR DESCRIPTION
**Pull request by gravityrail**
Fixes #7867

Some plans support different modules. This change declares plan support in the module header, and checks for that support when rendering the "active plan" information.

This allows us to easily render inactive rows in the `jetpack_modules` page and elsewhere, so that users know which modules are truly available.

This change also adds some basic checking in WP-cli and elsewhere for whether plan activation was successful.

#### Changes proposed in this Pull Request:

* Declare modules with a `Plan: business, premium, personal` (default is `free`) attribute
* Render full list of supported modules and plugins (in the case of vaultpress/akismet) in the `Jetpack::get_active_plan()` response
* Render rows as inactive in the `jetpack_modules` list if they are not supported by the current plan

The goal is NOT to prevent users from being able to hack the code to enable modules they shouldn't (since that is impossible anyway) but to help users understand what features their plan supports, and prevent them from accidentally enabling features that it doesn't.

#### Further work:

* Allow users to click to upgrade to a supporting plan from within the `jetpack_modules` page.

#### Testing instructions:

* On a site with a free plan:
  * Go to `/wp-admin/admin.php?page=jetpack_modules`
  * Ads, Data Backups, Google Analytics, SEO Tools, Search and VideoPress should be greyed out. Only Data Backups should be clickable, with a `Configure` link.
  * Go to `/settings/traffic/{site_domain}` in Calypso
    * Attempt to enable SEO tools using the `Enable` link next to `SEO Tools module is disabled in Jetpack.`
    * You should see a message `There was a problem saving your changes. Please try again.` (TODO: a better error message - in the network response you will see the actual error is `{error: "not_supported", message: "The requested Jetpack module is not supported by your plan."}`)
* On a site with a Premium plan
  * VideoPress should become available
* On a site with a Professional plan
  * SEO Tools, Google Analytics and Search should become available

#### Proposed changelog entry for your changes:

* Show correct `available` status in Jetpack modules list
* Return error in wp-cli if activating a module that is unsupported by your plan

### CODE REVIEW

File: class.jetpack-modules-list-table.php [hunk 1]
```diff
@@ -231,8 +220,10 @@ function single_row( $item ) {
 		if ( ! empty( $item['activated'] )  )
 			$row_class .= ' active';
 
-		if ( ! $this->is_module_available( $item ) )
+		if ( ! Jetpack_Admin::is_module_available( $item ) ) {
+			error_log($item['module']." is NOT available");
```
Comment (beaulebens) [inline class.jetpack-modules-list-table.php]: This shouldn't be here.

File: class.jetpack.php [hunk 2]
```diff
@@ -4033,7 +4051,9 @@ function admin_page_load() {
 				$module = stripslashes( $_GET['module'] );
 				check_admin_referer( "jetpack_activate-$module" );
 				Jetpack::log( 'activate', $module );
-				Jetpack::activate_module( $module );
+				if ( ! Jetpack::activate_module( $module ) ) {
+					Jetpack::state( 'error', sprintf( __( 'Could not activated %s', 'jetpack' ), $module ) );
```
Comment (beaulebens) [inline class.jetpack.php]: Typo: `Could not activate`

File: class.jetpack.php [hunk 3]
```diff
@@ -1485,12 +1493,8 @@ public static function get_active_plan() {
 		);
 
 		if ( in_array( $plan['product_slug'], $premium_plans ) ) {
-			$plan['supports'] = array(
-				'videopress',
-				'akismet',
-				'vaultpress',
-				'wordads',
-			);
```
Comment (beaulebens) [inline class.jetpack.php]: Did you mean to remove videopress and wordads from here?

File: class.jetpack.php [hunk 4]
```diff
@@ -1503,23 +1507,23 @@ public static function get_active_plan() {
...
 		if ( in_array( $plan['product_slug'], $business_plans ) ) {
-			$plan['supports'] = array(
-				'videopress',
-				'akismet',
-				'vaultpress',
-				'seo-tools',
-				'google-analytics',
-				'wordads',
-				'search',
-			);
```
Comment (beaulebens) [inline class.jetpack.php]: As above, did you mean to remove some items from this list?

File: class.jetpack.php (hunk 3 above)
Comment (gravityrail) [reply class.jetpack.php]: Yes - they are modules, the others are plugins. Modules are automatically added to the list based on metadata.

File: class.jetpack.php (hunk 4 above)
Comment (gravityrail) [reply class.jetpack.php]: As above - yes ;)

File: class.jetpack-modules-list-table.php (hunk 1 above)
Comment (gravityrail) [reply class.jetpack-modules-list-table.php]: Fixed

File: class.jetpack.php (hunk 2 above)
Comment (gravityrail) [reply class.jetpack.php]: Fixed

**Comment by beaulebens** [comment]
**On a Free plan**
Not a blocker, but noticed that all of the paid options are added to the bottom of the list, vs being in amongst the other items (in alpha order).

I don't have any configuration options on "Data Backups" (on Free plan), although maybe that's correct? I'm not sure what I'd actually configure at this point.

screen shot 2018-03-05 at 4 26 22 pm

I don't get the

> SEO Tools module is disabled in Jetpack.

banner in Calypso (correctly), so can't test that bit. I just get a banner asking me to upgrade to get access to that feature.

**On Personal**
No apparent differences compared to Free.

I'd expect Data Backups to be available, but it doesn't appear to be.

**On Premium**
Backups, Google Analytics, SEO and Search remain unavailable (VideoPress becomes available, as expected)

screen shot 2018-03-05 at 4 45 24 pm

**On Professional**
Everything was "available" _except_ "Data Backups". I _could_ see the "Configure" link when I hovered, but the module was still "disabled" in the list and could not be Activated via link or checkbox.

screen shot 2018-03-05 at 4 45 24 pm

**Comment by gravityrail** [comment]
@beaulebens - thanks for the detailed feedback.

> Not a blocker, but noticed that all of the paid options are added to the bottom of the list, vs being in amongst the other items (in alpha order).

That appears to be a deliberate feature that was added to the modules list at some point. By the way, the modules list is almost the worst piece of code I've seen in Jetpack. Let that sink in.

> I don't have any configuration options on "Data Backups" (on Free plan), although maybe that's correct? I'm not sure what I'd actually configure at this point.

> I'd expect Data Backups to be available, but it doesn't appear to be.

These are existing bugs unrelated to my changes, and should be handled in a separate PR. They also probably need some design/UX work to figure out how things need to behave.

Having now dealt with the code for this change, I actually recommend throwing away the whole modules list code and either just doing without it, or replacing it with a cleanroom implementation. In its current form it is very hard to understand what's going on. Case in point: The HTML implementation that is rendered out as a table appears to be completely replaced in the DOM by a Javascript implementation using JS templates at load time - all the data is being rendered twice, and it is impossible to tell why your changes are having no effect until you realise this is what's going on.

**Comment by gravityrail** [comment]
By the way, I'm not meaning to point fingers at anyone in particular about the code quality - this is also just about the OLDEST code in Jetpack, and it has just gone through many many many releases of being hacked and patched and never refactored. No surprise that it's got some serious warts.

**Comment by beaulebens** [comment]
Gotcha -- I hadn't realized that "Data Backups" was already in the list; I figured you'd added it as part of this. I've just tested `master` and am seeing it there as well, with all the same issues as mentioned above.

I've just gone through with that in mind and done some more testing, and this seems to handle things much better and more explicitly than the current `master`.

File: class.jetpack.php [hunk 5]
```diff
@@ -1452,6 +1452,13 @@ public static function refresh_active_plan_from_wpcom() {
 	 * @return array Active Jetpack plan details
 	 */
 	public static function get_active_plan() {
+		global $active_plan_cache;
```
Comment (beaulebens) [inline class.jetpack.php]: Is there any better way to do this than introducing another global?

File: class.jetpack.php [hunk 6]
```diff
@@ -1503,23 +1507,23 @@ public static function get_active_plan() {
...
+			$supports[] = 'akismet';
+			$supports[] = 'vaultpress';
 			$plan['class'] = 'business';
 		}
 
-		// Make sure we have an array here in the event database data is stale
-		if ( ! isset( $plan['supports'] ) ) {
-			$plan['supports'] = array();
+		// get available features
+		foreach( self::get_available_modules() as $module_slug ) {
```
Comment (beaulebens) [inline class.jetpack.php]: Spacing -- `foreach (`

File: class.jetpack.php (hunk 5 above)
Comment (gravityrail) [reply class.jetpack.php]: Unfortunately I couldn't think of one.

Comment (gravityrail) [reply class.jetpack.php]: To be clear: We could refactor the whole way that we list out modules, and then we'd only need to call this one - but that requires a lot more time than "hack week" necessarily allows.

**Comment by jeherve** [comment]
Related: #9033

**Pull request merged**